providers/mana/wr.c

```diff
@@ -0,0 +1,417 @@
+	case IBV_WR_SEND_WITH_INV:
+		flags |= GDMA_WORK_REQ_CHECK_SN;
+		send_oob.send.invalidate_key = wr->invalidate_rkey;
+		break;
+	case IBV_WR_SEND_WITH_IMM:
+		send_oob.send.immediate = be32toh(wr->imm_data);
+		SWITCH_FALLTHROUGH;
+	case IBV_WR_SEND:
+		flags |= GDMA_WORK_REQ_CHECK_SN;
+		break;
+	case IBV_WR_RDMA_WRITE:
+	case IBV_WR_RDMA_READ:
+		send_oob.rdma.address_hi = (uint32_t)(wr->wr.rdma.remote_addr >> 32);
+		send_oob.rdma.address_low = (uint32_t)(wr->wr.rdma.remote_addr & 0xFFFFFFFF);
+		send_oob.rdma.rkey = wr->wr.rdma.rkey;
+		send_oob.rdma.dma_len = msg_sz;
+		break;
+	default:
+		goto cleanup;
+	}
+
+	ret = gdma_post_sq_wqe(&qp->rc_qp.queues[USER_RC_SEND_QUEUE_REQUESTER], wr->sg_list,
+			       &send_oob, NULL, num_sge, MTU_SIZE(qp->mtu), flags, &gdma_wqe);
+	if (ret) {
+		verbs_err(verbs_get_ctx(qp->ibqp.qp.context),
+			  "rc post send error, ret %d\n", ret);
+		goto cleanup;
+	}
+
+	qp->rc_qp.acc_posted_wqe_size += gdma_wqe.size_in_bu;
+	qp->rc_qp.acc_read_posted_wqe_size += read_size_in_bu;
+	qp->rc_qp.sq_psn = PSN_ADD(qp->rc_qp.sq_psn, PSN_DELTA(msg_sz, qp->mtu));
+	qp->rc_qp.sq_ssn = PSN_INC(qp->rc_qp.sq_ssn);
+
+	*unmasked_wqe_index = gdma_wqe.unmasked_wqe_index;
+
+	return 0;
+
+cleanup:
+	return EINVAL;
+}
+
+static int mana_ib_rc_post_send(struct ibv_qp *ibqp, struct ibv_send_wr *wr,
+				struct ibv_send_wr **bad_wr)
+{
+	struct mana_qp *qp = container_of(ibqp, struct mana_qp, ibqp.qp);
+	uint32_t unmasked_queue_offset;
+	int ret;
+
+	pthread_spin_lock(&qp->sq_lock);
+
+	for (; wr; wr = wr->next) {
+		if ((wr->send_flags & IBV_SEND_SIGNALED) && shadow_queue_full(&qp->shadow_sq)) {
+			verbs_err(verbs_get_ctx(ibqp->context), "shadow queue full\n");
+			ret = ENOMEM;
+			goto cleanup;
+		}
+
+		if (!is_opcode_supported(wr->opcode)) {
+			ret = EINVAL;
+			goto cleanup;
+		}
+
+		ret = mana_ib_rc_post_send_request(qp, wr, &unmasked_queue_offset);
+		if (ret) {
+			verbs_err(verbs_get_ctx(qp->ibqp.qp.context),
+				  "failed to post send request ret %d\n", ret);
+			goto cleanup;
+		}
+
+		if (wr->send_flags & IBV_SEND_SIGNALED) {
+			/* Fill shadow queue data */
+			struct rc_sq_shadow_wqe *shadow_wqe = (struct rc_sq_shadow_wqe *)
+				shadow_queue_producer_entry(&qp->shadow_sq);
+			memset(shadow_wqe, 0, sizeof(struct rc_sq_shadow_wqe));
+			shadow_wqe->header.wr_id = wr->wr_id;
+			shadow_wqe->header.opcode = convert_wr_to_wc(wr->opcode);
+			shadow_wqe->header.posted_wqe_size_in_bu = qp->rc_qp.acc_posted_wqe_size;
+			shadow_wqe->header.unmasked_queue_offset = unmasked_queue_offset;
+			shadow_wqe->end_psn = PSN_DEC(qp->rc_qp.sq_psn);
+			shadow_wqe->read_posted_wqe_size_in_bu =
+				qp->rc_qp.acc_read_posted_wqe_size;
+			shadow_queue_advance_producer(&qp->shadow_sq);
+
+			qp->rc_qp.acc_read_posted_wqe_size = 0;
+			qp->rc_qp.acc_posted_wqe_size = 0;
+		}
+
+		mana_ib_update_shared_mem_right_offset(qp, unmasked_queue_offset);
+		gdma_ring_send_doorbell(&qp->rc_qp.queues[USER_RC_SEND_QUEUE_REQUESTER]);
```

Comment:
Is it necessary to ring the doorbell once per wr within the loop body, or can the doorbell ring be moved outside the loop to batch multiple submissions?